pennylane/tape/tapes/jacobian_tape.py

```diff
@@ -239,22 +239,69 @@ def numeric_pd(self, idx, device, params=None, **options):
         shift[idx] = h
 
         if order == 1:
-            # Forward finite-difference.
-            # Check if the device has already be pre-computed with
-            # unshifted parameter values, to avoid redundant evaluations.
+            # forward finite-difference.
+
+            tapes = []
+
+            # get the stored result of the original circuit
             y0 = options.get("y0", None)
 
+            shifted = self.copy(deep=True, tape_cls=QuantumTape)
+            shifted.set_parameters(params + shift)
+
+            tapes.append(shifted)
+
             if y0 is None:
-                y0 = np.asarray(self.execute_device(params, device))
+                tapes.append(self)
 
-            y = np.asarray(self.execute_device(params + shift, device))
-            return (y - y0) / h
+            def processing_fn(results):
+                """Computes the gradient of the parameter at index idx via first-order
+                forward finite differences.
+
+                Args:
+                    results (list[real]): evaluated quantum tapes
+
+                Returns:
+                    array[float]: 1-dimensional array of length determined by the tape output
+                    measurement statistics
+                """
+                shifted = np.array(results[0])
+                unshifted = y0
+
+                if unshifted is None:
+                    unshifted = np.array(results[1])
+
+                return (shifted - unshifted) / h
+
+            return tapes, processing_fn
 
         if order == 2:
             # central finite difference
-            shift_forward = np.array(self.execute_device(params + shift / 2, device))
-            shift_backward = np.array(self.execute_device(params - shift / 2, device))
-            return (shift_forward - shift_backward) / h
+
+            shifted_forward = self.copy(deep=True, tape_cls=QuantumTape)
+            shifted_forward.set_parameters(params + shift / 2)
+
+            shifted_backward = self.copy(deep=True, tape_cls=QuantumTape)
```

Comment:
Hmm, the problem with having a single tape and just changing the parameters is that this kind of logic would require something like 

```
original_params = tape.get_parameters()
tape.set_parameters(shifted_params)
tape.execute()
tape.set_parameters(original_params)
```

which would now have to happen inside the device, which needs to know when to invoke this behaviour. I don't see how this can happen in the new (otherwise extremely clean) logic.


The reason why I am not to worried about copy times and memory issues is that our vision is to use an efficient single-tape representation of multiple tapes. So the copying would be replaced by adding branches to this single tape.

Reply:
Note that these aren't deep copies! The keyword argument name is deceptive, I'll change it. This is just (two) shallow copies.

Reply:
@mariaschuld yep understood. Just wanted to make sure that even with the shorter-term solution we are not unnecessarily introducing possible bottlenecks